### ISSUE TITLE
Implementation Details of data2vec

### ISSUE THREAD

**Issue by cashincashout** [facebookresearch/fairseq#4199]
Hi @alexeib,

I'm trying to reproduce the results of data2vec (vision part) but got a few questions about the implementation details.
1. Is any projection head after the student/teacher network being used? If yes, what is the detailed architecture?
2. The paper mentions using the layer normalization of output after FFN of each block as the target. ViT has a default `self.norm` at the final layer for classification. Are the layer norms you use here sharing the parameters with this default `self.norm`?

Best

**Comment by alexeib** [comment]
1. yes there is a projection head, it is a single linear layer (with bias)
2. no, you need to use parameter-free normalization so that the model doesnt learn to favor a particular layer over others

**Comment by jizxny** [comment]
@alexeib,

Oh, I see. I note that in the paper you write "parameter-less layer normalization" instead of "parameter-free". So here did you use `nn.LayerNorm(embed_dim, elementwise_affine=False)` here?

**Comment by shallowtoil** [comment]
Hi @alexeib. Thanks for the clarification. Also, did you use weight decay with Adam optimizer?

**Comment by cashincashout** [comment]
Hi @alexeib. Thanks for the clarification. Also, did you use weight decay?

**Comment by cashincashout** [comment]
I am trying to reproduce the results using the [dino](https://github.com/facebookresearch/dino) repo. I closely follow the implementation details in the paper, but the loss decreases to 0 at early epochs (at ~30 epoch).  You can check out here the [training args and log](https://drive.google.com/file/d/1kICEL1WrplxRfpcwa2tukuSdUaexKIbM/view?usp=sharing) (see patch loss). Do you have any related experience in this situation? Thanks for your help!

**Comment by carrascomj** [comment]
According to the paper, you need to use a warmup temperature (0.002) for some epochs (40 and 80 in the paper) for the learning rate. That's highlighted as important to avoid collapse. In the dino repo, there is a cosine scheduler for the learning rate  where you can put the warmup parameters (also through the CLI).

I am not sure whether the warmup learning rate is actually used as in dino; that is, a range between the warmup value and the starting value of the cosine schedule:

```python
np.linspace(warmup_lr, start_learning_rate, warmup_epochs)
```

or whether it is just a fixed value (or if it matters). I would say it's a fixed value, since they did not say the opposite.

Also, the stochastic depth may be important to avoid the collapse. In dino's repository, it is 0.1 by default and applied to the student (0.2 in data2vec).
